src/dbnode/storage/shard.go

```diff
@@ -2078,6 +2159,17 @@ func (s *dbShard) Repair(
 	return repairer.Repair(ctx, s.namespace.ID(), tr, s)
 }
 
+func (s *dbShard) TagsFromSeriesID(seriesID ident.ID) (ident.Tags, error) {
+	s.RLock()
+	entry, _, err := s.lookupEntryWithLock(seriesID)
+	s.RUnlock()
+	if entry == nil || err != nil {
```

Comment:
I think `entry == nil` should be handled separately. I.E you probably want to propagate that information to the caller (with an `ok bool` or something) so that they can make a decision (in your case, probably to skip since it clearly fell out of retention or something)

Reply:
Also technically its fine right now but to be extra safe can you change `newShardEntry` in `shard.go` to mark all of the tags as `NoFinalize()` to ensure that they never get put back in a pool while the merge logic still has a handle on them?